Android App Crashes while sending message to users/ on channel [security bug]

## Description
 I found a security vulnerability in Rocket's latest android app by which I was able to remotely crash any  user’s app  instantly just by just sending a simple message in private or in channel. The vulnerability  require the victim open the message. 

<!-- Please, describe what's the issue here. -->

## Devices and Versions

<!-- Version can be found by opening the side menu and then clicking on "Settings" and then "About" -->
Rocket.Chat.Android version: (e.g. 4.5.1)

<!-- Found a bug? List all devices that reproduced it and all that doesn't -->
Mobile device model and OS version: (tested on :+1: -- " **Android 6.0, 8.0, 10.0**"), probably any other android version

## Steps to reproduce

> Create new #test channel
> Send POC Code onto the channel
> Open Mobile App
> App gets crashed

<!-- In case it is a bug, can you describe the steps to reproduce it please? -->

## POC
### Crafted code to crash mobile app
![](https://i.postimg.cc/zvBWdMzT/Screenshot-20200320-112405.png)

### Message Preview
![](https://i.postimg.cc/fbCJ6KgC/Screenshot-20200320-112541.png)

### App Gets Crashed
![](https://i.postimg.cc/26J8DXdQ/Screenshot-20200320-112711.png)

### Code Link
https://pastebin.com/raw/JEDcC5Yr

**There is no such problem in iOS client and rocket web**

I was able to reproduce it.
Error:
```
03-20 08:43:56.892 19002 19471 E AndroidRuntime: FATAL EXCEPTION: OkHttp Dispatcher
03-20 08:43:56.892 19002 19471 E AndroidRuntime: Process: chat.rocket.android, PID: 19002
03-20 08:43:56.892 19002 19471 E AndroidRuntime: java.lang.RuntimeException: java.net.URISyntaxException: Illegal character in authority at index 8: https://google.com+`{}alph4`/
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at okhttp3.HttpUrl.uri(HttpUrl.java:386)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at okhttp3.internal.connection.RouteSelector.resetNextProxy(RouteSelector.java:129)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at okhttp3.internal.connection.RouteSelector.<init>(RouteSelector.java:63)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at okhttp3.internal.connection.StreamAllocation.<init>(StreamAllocation.java:101)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:112)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:254)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:200)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at java.lang.Thread.run(Thread.java:919)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: Caused by: java.net.URISyntaxException: Illegal character in authority at index 8: https://google.com+`{}alph4`/
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at java.net.URI$Parser.fail(URI.java:2893)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at java.net.URI$Parser.parseAuthority(URI.java:3231)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at java.net.URI$Parser.parseHierarchical(URI.java:3142)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at java.net.URI$Parser.parse(URI.java:3098)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at java.net.URI.<init>(URI.java:584)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	at okhttp3.HttpUrl.uri(HttpUrl.java:379)
03-20 08:43:56.892 19002 19471 E AndroidRuntime: 	... 12 more
```

--- a/app/containers/markdown/index.js
+++ b/app/containers/markdown/index.js
@@ -278,7 +278,7 @@ class Markdown extends PureComponent {
 		);
 	}
 
-	renderImage = ({ src }) => <Image style={styles.inlineImage} source={{ uri: src }} />;
+	renderImage = ({ src }) => <Image style={styles.inlineImage} source={{ uri: encodeURI(src) }} />;
 
 	renderEditedIndicator = () => {
 		const { theme } = this.props;

--- a/app/containers/markdown/index.js
+++ b/app/containers/markdown/index.js
@@ -22,6 +22,7 @@ import MarkdownTableCell from './TableCell';
 import mergeTextNodes from './mergeTextNodes';
 
 import styles from './styles';
+import { isValidURL } from '../../utils/url';
 
 // Support <http://link|Text>
 const formatText = text => text.replace(
@@ -278,7 +279,18 @@ class Markdown extends PureComponent {
 		);
 	}
 
-	renderImage = ({ src }) => <Image style={styles.inlineImage} source={{ uri: encodeURI(src) }} />;
+	renderImage = ({ src }) => {
+		if (!isValidURL(src)) {
+			return null;
+		}
+
+		return (
+			<Image
+				style={styles.inlineImage}
+				source={{ uri: encodeURI(src) }}
+			/>
+		);
+	}
 
 	renderEditedIndicator = () => {
 		const { theme } = this.props;

--- a/app/utils/url.js
+++ b/app/utils/url.js
@@ -1,0 +1,9 @@
+export const isValidURL = (url) => {
+	const pattern = new RegExp('^(https?:\\/\\/)?' // protocol
+		+ '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|' // domain name
+		+ '((\\d{1,3}\\.){3}\\d{1,3}))' // OR ip (v4) address
+		+ '(\\:\\d+)?(\\/[-a-z\\d%_.~+]*)*' // port and path
+		+ '(\\?[;&a-z\\d%_.~+=-]*)?' // query string
+		+ '(\\#[-a-z\\d_]*)?$', 'i'); // fragment locator
+	return !!pattern.test(url);
+};

